commands/src/main/java/org/wildfly/extras/creaper/commands/deployments/Deploy.java

```diff
@@ -0,0 +1,211 @@
+        ModelNode content = new ModelNode();
+        content.get(Constants.INPUT_STREAM_INDEX).set(0);
+
+        ModelNode deployRequest;
+
+        if (ctx.options.isDomain) {
+            List<String> groups = serverGroups;
+            if (serverGroups == null)
+                groups = ops.readChildrenNames(Address.root(), Constants.SERVER_GROUP)
+                        .stringListValue(Collections.<String>emptyList());
+
+            List<ModelNode> groupDeploymentAddresses = new ArrayList<ModelNode>(groups.size());
+            for (String serverGroup : groups) {
+                ModelNode mn = new ModelNode();
+                mn.add(Constants.SERVER_GROUP, serverGroup);
+                mn.add(Constants.DEPLOYMENT, deploymentName);
+                groupDeploymentAddresses.add(mn);
+            }
+            deployRequest = createDeploymentOperation(content,
+                    groupDeploymentAddresses.toArray(new ModelNode[groupDeploymentAddresses.size()]));
+        } else {
+            deployRequest = createDeploymentOperation(content);
+            ModelNode deployStep = deployRequest.get(Constants.STEPS).add();
+            ModelNode deploymentAddress = new ModelNode();
+            deploymentAddress.add(Constants.DEPLOYMENT, deploymentName);
+            deployStep.set(getEmptyOperation(DEPLOY, deploymentAddress));
+        }
+
+        OperationBuilder builder = new OperationBuilder(deployRequest, autoCloseInputStream);
+        builder.addInputStream(deploymentInputStream);
+        ctx.client.execute(builder.build());
+    }
+
+    private ModelNode createDeploymentOperation(ModelNode content, ModelNode... serverGroupAddresses) {
+        ModelNode composite = getEmptyOperation(Constants.COMPOSITE, new ModelNode().setEmptyList());
+        ModelNode steps = composite.get(Constants.STEPS);
+        ModelNode step1 = steps.add();
+        step1.set(getEmptyOperation(Constants.ADD, new ModelNode().add(Constants.DEPLOYMENT, deploymentName)));
+        step1.get(Constants.CONTENT).add(content);
+        if (serverGroupAddresses != null) {
+            for (ModelNode serverGroup : serverGroupAddresses) {
+                ModelNode sg = steps.add();
+                sg.set(getEmptyOperation(Constants.ADD, serverGroup));
+                sg.get(ENABLED).set(true);
+            }
+        }
+        return composite;
+    }
+
+    private ModelNode getEmptyOperation(String operationName, ModelNode address) {
+        ModelNode op = new ModelNode();
+        op.get(Constants.OP).set(operationName);
+        if (address != null) {
+            op.get(Constants.OP_ADDR).set(address);
+        } else {
+            // Just establish the standard structure; caller can fill in address later
+            op.get(Constants.OP_ADDR);
+        }
+        return op;
+    }
+
+    /**
+     * A single shared instance of this class is passed through the entire builder invocation chain to accumulate
+     * all the options.
+     */
```

Comment:
This is clearly copied from `Online|OfflineOptions`, where I deliberately broke the builder to a bunch of classes to be able to provide compile-time checking of all mandatory options. It adds quite some maintenance burden, so I don't think it's suitable for commands. Did you have any special reason to use this pattern instead of plain old single-class builder?